File: src/python/beancount/core/account.py
"""Functions that operate on account strings.

These account objects are rather simple and dumb; they do not contain the list
of their associated postings. This is achieved by building a realization; see
realization.py for details.
"""
__author__ = "Martin Blais <blais@furius.ca>"

import re
import os
from os import path


# Component separator for account names.
# pylint: disable=invalid-name
sep = ':'


# Regular expression string that matchs a valid account.
ACCOUNT_RE = '[A-Z][A-Za-z0-9\-]+(?:{}[A-Z][A-Za-z0-9\-]*)+'.format(sep)


# A dummy object which stands for the account type. Values in custom directives
# use this to disambiguate between string objects and account names.
TYPE = '<AccountDummy>'


def is_valid(string):
    """Return true if the given string is a valid account name.
    This does not check for the root account types, just the general syntax.

    Args:
      string: A string, to be checked for account name pattern.
    Returns:
      A boolean, true if the string has the form of an account's name.
    """
    return (isinstance(string, str) and
            bool(re.match('{}$'.format(ACCOUNT_RE), string)))


def join(*components):
    """Join the names with the account separator.

    Args:
      *components: Strings, the components of an account name.
    Returns:
      A string, joined in a single account name.
    """
    return sep.join(components)


def split(account_name):
    """Split an account's name into its components.

    Args:
      account_name: A string, an account name.
    Returns:
      A list of strings, the components of the account name (without the separators).
    """
    return account_name.split(sep)


def parent(account_name):
    """Return the name of the parent account of the given account.

    Args:
      account_name: A string, the name of the account whose parent to return.
    Returns:
      A string, the name of the parent account of this account.
    """
    assert isinstance(account_name, str), account_name
    if not account_name:
        return None
    components = account_name.split(sep)
    components.pop(-1)
    return sep.join(components)


def leaf(account_name):
    """Get the name of the leaf of this account.

    Args:
      account_name: A string, the name of the account whose leaf name to return.
    Returns:
      A string, the name of the leaf of the account.
    """
    assert isinstance(account_name, str)
    return account_name.split(sep)[-1] if account_name else None


def sans_root(account_name):
    """Get the name of the account without the root.

    For example, an in put of 'Assets:BofA:Checking' will produce 'BofA:Checking'.

    Args:
      account_name: A string, the name of the account whose leaf name to return.
    Returns:
      A string, the name of the non-root portion of this account name.
    """
    assert isinstance(account_name, str)
    components = account_name.split(sep)[1:]
    return join(*components) if account_name else None


def root(num_components, account_name):
    """Return the first few components of an account's name.

    Args:
      num_components: An integer, the number of components to return.
      account_name: A string, an account name.
    Returns:
      A string, the account root up to 'num_components' components.
    """
    return join(*(split(account_name)[:num_components]))


def has_component(account_name, component):
    """Return true if one of the account contains a given component.

    Args:
      account_name: A string, an account name.
      component: A string, a component of an account name. For instance,
        'Food' in 'Expenses:Food:Restaurant'. All components are considered.
    Returns:
      A boolean, true if the component is in the account. Note that a component
      name must be whole, that is 'NY' is not in Expenses:Taxes:StateNY'.
    """
    return re.search('(^|:){}(:|$)'.format(component), account_name)


def commonprefix(accounts):
    """Return the common prefix of a list of account names.

    Args:
      accounts: A sequence of account name strings.
    Returns:
      A string, the common parent account. If none, returns an empty string.
    """
    accounts_lists = [account_.split(sep)
                      for account_ in accounts]
    # Note: the os.path.commonprefix() function just happens to work here.
    # Inspect its code, and even the special case of no common prefix
    # works well with str.join() below.
    common_list = path.commonprefix(accounts_lists)
    return sep.join(common_list)


def walk(root_directory):
    """A version of os.walk() which yields directories that are valid account names.

    This only yields directories that are accounts... it skips the other ones.
    For convenience, it also yields you the account's name.

    Args:
      root_directory: A string, the name of the root of the hierarchy to be walked.
    Yields:
      Tuples of (root, account-name, dirs, files), similar to os.walk().
    """
    for root, dirs, files in os.walk(root_directory):
        dirs.sort()
        files.sort()
        relroot = root[len(root_directory)+1:]
        account_name = relroot.replace(os.sep, sep)
        if is_valid(account_name):
            yield (root, account_name, dirs, files)


def parent_matcher(account_name):
    """Build a predicate that returns whether an account is under the given one.

    Args:
      account_name: The name of the parent account we want to check for.
    Returns:
      A callable, which, when called, will return true if the given account is a
      child of 'account_name'.
    """
    return re.compile(r'{}\b'.format(re.escape(account_name))).match


def parents(account_name):
    """A generator of the names of the parents of this account, including this account.

    Args:
      account_name: The name of the account we want to start iterating from.
    Returns:
      A generator of account name strings.
    """
    while account_name:
        yield account_name
        account_name = parent(account_name)

File: experiments/golang/core/account.go
package core

import (
	"regexp"
	"strings"
)

// Account separator for account names.
const AccountSeparator = ":"

// Regular expressions for account validation.
// Beancount account components must start with an uppercase letter, a digit, or any non-ASCII character.
// UTF-8-ONLY in Beancount is defined as any multi-byte UTF-8 character.
// In Go, we can use [^\x00-\x7f] to match any non-ASCII character.
const (
	utf8Only      = `[^\x00-\x7f]`
	accCompTypeRE = `(?:[A-Z]|` + utf8Only + `)(?:[A-Za-z0-9\-]|` + utf8Only + `)*`
	accCompNameRE = `(?:[A-Z0-9]|` + utf8Only + `)(?:[A-Za-z0-9\-]|` + utf8Only + `)*`
)

var (
	validRootRe = regexp.MustCompile("^" + accCompTypeRE + "$")
	validAccountRe = regexp.MustCompile(`^(?:` + accCompTypeRE + `)(?:` + AccountSeparator + accCompNameRE + `)+$`)
)

// IsValidRoot returns true if the string is a valid root account name.
func IsValidRoot(s string) bool {
	return validRootRe.MatchString(s)
}

// IsValid returns true if the string is a valid account name.
func IsValid(s string) bool {
	return validAccountRe.MatchString(s)
}

// Join joins the components with the account separator.
func Join(components ...string) string {
	return strings.Join(components, AccountSeparator)
}

// Split splits an account's name into its components.
func Split(accountName string) []string {
	if accountName == "" {
		return nil
	}
	return strings.Split(accountName, AccountSeparator)
}

// Parent returns the name of the parent account or an empty string if at the root.
func Parent(accountName string) string {
	if accountName == "" {
		return ""
	}
	idx := strings.LastIndex(accountName, AccountSeparator)
	if idx == -1 {
		return ""
	}
	return accountName[:idx]
}

// Leaf returns the name of the leaf component of the account.
func Leaf(accountName string) string {
	if accountName == "" {
		return ""
	}
	idx := strings.LastIndex(accountName, AccountSeparator)
	if idx == -1 {
		return accountName
	}
	return accountName[idx+len(AccountSeparator):]
}

// Parents returns all parent accounts including the account itself.
func Parents(accountName string) []string {
	if accountName == "" {
		return nil
	}
	components := Split(accountName)
	res := make([]string, len(components))
	for i := range components {
		res[i] = Join(components[:i+1]...)
	}
	// Reverse to match Python's parents (child to root)
	for i, j := 0, len(res)-1; i < j; i, j = i+1, j-1 {
		res[i], res[j] = res[j], res[i]
	}
	return res
}

// HasComponent returns true if the account contains a given component.
func HasComponent(accountName, component string) bool {
	components := Split(accountName)
	for _, c := range components {
		if c == component {
			return true
		}
	}
	return false
}

// SansRoot returns the account name without its root component.
func SansRoot(accountName string) string {
	components := Split(accountName)
	if len(components) <= 1 {
		return ""
	}
	return Join(components[1:]...)
}

// Root returns the root component of the account name up to a given depth.
func Root(depth int, accountName string) string {
	components := Split(accountName)
	if len(components) == 0 {
		return ""
	}
	if depth > len(components) {
		depth = len(components)
	}
	if depth <= 0 {
		return ""
	}
	return Join(components[:depth]...)
}

// CommonPrefix returns the common account prefix between multiple accounts.
func CommonPrefix(accounts []string) string {
	if len(accounts) == 0 {
		return ""
	}
	common := Split(accounts[0])
	for _, acc := range accounts[1:] {
		comp := Split(acc)
		n := len(common)
		if len(comp) < n {
			n = len(comp)
		}
		newCommon := []string{}
		for i := 0; i < n; i++ {
			if common[i] == comp[i] {
				newCommon = append(newCommon, common[i])
			} else {
				break
			}
		}
		common = newCommon
	}
	return Join(common...)
}

// ParentMatcher returns a function that matches an account and its subaccounts.
func ParentMatcher(parent string) func(string) bool {
	prefix := parent + AccountSeparator
	return func(account string) bool {
		return account == parent || strings.HasPrefix(account, prefix)
	}
}

// AccountTransformer handles renaming or transforming account names.
type AccountTransformer struct {
	rsep string
}

func NewAccountTransformer(rsep string) *AccountTransformer {
	return &AccountTransformer{rsep: rsep}
}

func (t *AccountTransformer) Render(accountName string) string {
	if t.rsep == "" {
		return accountName
	}
	return strings.ReplaceAll(accountName, AccountSeparator, t.rsep)
}

func (t *AccountTransformer) Parse(transformedName string) string {
	if t.rsep == "" {
		return transformedName
	}
	return strings.ReplaceAll(transformedName, t.rsep, AccountSeparator)
}
